packages/astro/src/content/mutable-data-store.ts

```diff
@@ -190,6 +192,29 @@ export default new Map([\n${lines.join(',\n')}]);
 		}
 	}
 
+	#writing = new Set<string>();
+	#pending = new Set<string>();
+
+	async #writeFileAtomic(filePath: PathLike, data: string, depth = 0) {
+		const fileKey = filePath.toString();
+		if (this.#writing.has(fileKey)) {
+			this.#pending.add(fileKey);
+			return;
+		}
+		const tempFile = filePath instanceof URL ? new URL(`${filePath.href}.tmp`) : `${filePath}.tmp`;
+		this.#writing.add(fileKey);
+		try {
+			await fs.writeFile(tempFile, data);
+			await fs.rename(tempFile, filePath);
+		} finally {
+			this.#writing.delete(fileKey);
+			if (this.#pending.has(fileKey) && depth < MAX_DEPTH) {
+				this.#pending.delete(fileKey);
+				await this.#writeFileAtomic(filePath, data, depth + 1);
+			}
```

Comment:
I am not really sure what this code does and why we use the resursion, but would happen if the pending files exceed the `	MAX_DEPTH`?

Reply:
The depth isn't the number of files: it's the number of times that it has finished writing a file and found that something has tried to write it again. I wasn't able to make it go beyond 1 in my tests, but I put that in as a safety to avoid the scenario where something is constantly trying to write and end up makign a massively deep call stack. All it would do would skip that one write, and then presumably it will try to write again immediatley after.

Reply:
Sorry, I needed to be clear. Here's a couple of things:
- the business logic of the function isn't clear, so it's best to write a comment that explains, especially why
- the check of the depth is done with the pending writes, it isn't very clear why and how. Since the correlation is hard to grasp (probably because I don't understand the business logic), my question was mostly regarding the code: the pending list is incremented at the beginning of the function, and it doesn't depend on the depth; however the deletion of items from the list is conditional to the depth. My question is: don't we need to flush them? If the list has some items, is it a bug? 

Apologies for my question might sound dumb, I'm just trying to understand what's on.

Reply:
OK, I've moved things around a bit and added lots of comments to hopefully make it clearer